File: eileenjang/LeetCode/417_Pacific_Atlantic_Water_Flow.java
class Solution {
    int dir[][] = {{0,1}, {0,-1}, {1,0}, {-1,0}};
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        List<List<Integer>> result = new ArrayList<>();
        if(heights.length == 0 || heights[0].length == 0 || heights == null) 
            return result;
        
        int row = heights.length, col = heights[0].length;
        boolean[][] pacific = new boolean[row][col];
        boolean[][] atlantic = new boolean[row][col];
        
        // Dynamic Frequency Selection
        for(int i = 0; i < col; i++){
            dfs(heights, 0, i, Integer.MIN_VALUE, pacific);
            dfs(heights, row-1, i, Integer.MIN_VALUE, atlantic);
        }
        for(int i = 0; i < row; i++){
            dfs(heights, i, 0, Integer.MIN_VALUE, pacific);
            dfs(heights, i, col-1, Integer.MIN_VALUE, atlantic);
        }
        
        // Before making result
        for(int i = 0; i < row; i++){
            for(int j = 0; j < col; j++) {
                if(pacific[i][j] && atlantic[i][j]) {
                    result.add(Arrays.asList(i,j));
                }
            }
        }
        
        return result;
    }
    
    public void dfs(int[][] heights, int i, int j, int prev, boolean[][] ocean){
        if(i < 0 || i >= ocean.length || j < 0 || j >= ocean[0].length) return;
        if(heights[i][j] < prev || ocean[i][j]) return;
        ocean[i][j] = true;
        for(int[] d : dir){
            dfs(heights, i+d[0], j+d[1], heights[i][j], ocean);
        }
        
    }
}

// TC: O(M*N); SC: O(M*N)

File: paxtaeo/LeetCode/417_Pacific_Atlantic_Water_Flow.py
class Solution:
    def dfs(self, heights, r, c, visited):
        m, n = len(heights), len(heights[0])
        directions = ((0, 1), (0, -1), (1, 0), (-1, 0))

        visited[r][c] = True
        for d in directions:
            nr, nc = r+d[0], c+d[1]
            if nr < 0 or nr >= m or nc < 0 or nc >= n or visited[nr][nc] or heights[r][c] > heights[nr][nc]:
                continue
            self.dfs(heights, nr, nc, visited)

    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if len(heights) == 0:
            return []
        
        m, n = len(heights), len(heights[0])
        visited_p = [[False for _ in range(n)] for _ in range(m)]
        visited_a = [[False for _ in range(n)] for _ in range(m)]

        for cr in range(m):
            self.dfs(heights, cr, 0, visited_p)
            self.dfs(heights, cr, n-1, visited_a)
        
        for cc in range(n):
            self.dfs(heights, 0, cc, visited_p)
            self.dfs(heights, m-1, cc, visited_a)

        ans = []
        for r in range(m):
            for c in range(n):
                if visited_p[r][c] and visited_a[r][c]:
                    ans.append([r, c])
        
        return ans